Redesign: Change copy icon.

nimishmedatwal opened the pull request:
Replaces current "copy" icon with the Feather "copy" icon everywhere that it's used along with adding it's attribution in  /docs/thirdparty.md.

<!-- Describe your pull request here.-->

Fixes: #28700

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/91622060/cf72f1b4-df4f-425f-905b-3001925a455a)

![Screenshot 2024-03-24 164829](https://github.com/zulip/zulip/assets/91622060/74c9ef39-ce86-4639-a70a-824fe88f3f83)
![Screenshot 2024-03-24 164803](https://github.com/zulip/zulip/assets/91622060/511e24f7-47e2-4378-97f6-4ea8cd2c8417)

![Screenshot 2024-03-24 164816](https://github.com/zulip/zulip/assets/91622060/a366fffd-c1fe-4ccc-97b1-9d8d96a2151b)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Hm, the tooltip doesn't look centered over the "Copy code" icon, and the icon size/alignment doesn't look good to me in the first two screenshots.

alya commented:
I'm not sure what exactly is off about it. Including one-to-one before/after comparisons would be helpful; see @karlstolley 's PRs for some examples.

nimishmedatwal commented:
> I'm not sure what exactly is off about it. Including one-to-one before/after comparisons would be helpful; see @karlstolley 's PRs for some examples of how to do that.

Sure I'll do that !

karlstolley commented on docs/THIRDPARTY:
```diff
@@ -1472,3 +1476,22 @@ License: GPL-2.0+
  On Debian systems, the complete text of the GNU General Public
  License can be found in /usr/share/common-licenses/GPL-2 file.
+
+License: The MIT License (MIT)
```
This license is covered by Expat. There's no need to include it here.

karlstolley commented on docs/THIRDPARTY:
```diff
@@ -59,6 +59,10 @@ Files: static/generated/emoji/images/emoji/unicode/*
 
+Files: static/images/help/copy_to_clipboard.svg
+Copyright: 2013-2023 Cole Bemis
+License: The MIT License (MIT)
```
List this as Expat. See other examples in `docs/THIRDPARTY`

karlstolley commented on web/templates/copy_message_button.hbs:
```diff
@@ -1,3 +1,5 @@
 <div class="copy_button_base copy_message tippy-zulip-tooltip" data-tippy-content="{{t 'Copy and close' }}" aria-label="{{t 'Copy and close' }}" role="button">
-    {{> copy_to_clipboard_svg }}
```
What in the world is going on with these templates files? I don't understand why you're duplicating things here, or why you're copying and pasting the contents of `copy_to_clipboard.svg`

nimishmedatwal replied:
This was something I was experimenting with and got into the pr by mistake, I've fixed it now.

nimishmedatwal commented:
@karlstolley I've changed a lot of things in this pr, can you please give it another review?

nimishmedatwal commented on web/templates/copy_to_clipboard_svg.hbs:
```diff
@@ -1,3 +1,3 @@
-<svg height="20" width="16" viewBox="0 0 1000 1000" xmlns="http://www.w3.org/2000/svg" id="clipboard_image">
-    <path fill="#777" d="M128 768h256v64H128v-64z m320-384H128v64h320v-64z m128 192V448L384 640l192 192V704h320V576H576z m-288-64H128v64h160v-64zM128 704h160v-64H128v64z m576 64h64v128c-1 18-7 33-19 45s-27 18-45 19H64c-35 0-64-29-64-64V192c0-35 29-64 64-64h192C256 57 313 0 384 0s128 57 128 128h192c35 0 64 29 64 64v320h-64V320H64v576h640V768zM128 256h512c0-35-29-64-64-64h-64c-35 0-64-29-64-64s-29- ...
+<svg width="16" height="16" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg" class="copy_to_clipboard">
```
@karlstolley I did it since that was the way it was done earlier but yeah I'll change that.

nimishmedatwal commented:
@karlstolley I've changed `copy_to_clipboard_svg.hbs` so that it now uses <img> tag instead like you suggested.

One more thing about copy button that bugs me is that when you view a message source and message is quite long then the copy icon and the message overlaps which makes it difficult to see both.
![image](https://github.com/zulip/zulip/assets/91622060/817f49da-ed2e-4d3a-ad84-13731e1a0fbb)

karlstolley commented:
That is annoying. See #29247 for the blurred-background solution that was merged for buttons on code blocks. You might look into replicating that in a follow-up PR for the message area, probably with a bit of design discussion on CZO first.

karlstolley commented:
All right. Please re-create all of your screenshots with the original PR above. The very first one I checked is all wrong:

![Screenshot 2024-04-02 at 2 33 23 PM](https://github.com/zulip/zulip/assets/170719/494227ca-d1f6-4b9e-9a30-91d487567437)

nimishmedatwal commented:
One bug I noticed while testing is that if a user accidently clicks and drags the copy icon then upon releasing the icon will get uploaded as an .svg file in compose. This might be because of putting the svg in <img> tag. Now this can be fixed by setting `user-drag: none;` but I am sure there are better ways to fix this than this. 

https://github.com/zulip/zulip/assets/91622060/c576583b-caac-4a99-bda5-34ffcd89665f

nimishmedatwal commented:
> I'm still puzzled by the copy-paste job here. Would it be possible to just reference the `copy_to_clipboard.svg` file here in an `<img>` tag? There might be side-effects with that, but ideally I would love to see the SVG asset as loosely coupled with the template file here as possible.

@karlstolley the major problem I am facing after shifting to `<img>`  tag is that on hover I am not able to change the color of svg.

karlstolley commented:
Ah, that's right. Okay, so how about this plan: your `copy_to_clipboard_svg.hbs` should get the `<svg>` portion of this in its pure form, with any additional variables (IDs, classes) passable as arguments to the template, rendering the other `copy_xxx.hbs` files unnecessary (ideally). 

Get rid of the unused `.svg` file, I think, and point the license info to the `copy_to_clipboard_svg.hbs` file entirely.

nimishmedatwal commented:
@karlstolley I think I've incorporated all the changes that you suggested, it is working perfectly while testing (change of icon's color on hover is also working). I've also deleted `copy_xxx.hbs` files. Let me know if you have any other suggestion.

karlstolley commented:
Something is not right. I'm still seeing this on Firefox/Mac:

![Screenshot 2024-05-01 at 2 33 22 PM](https://github.com/zulip/zulip/assets/170719/380c1cc9-2c1c-415b-9a09-37121c8d48a2)

And Chrome/Mac:
![Screenshot 2024-05-01 at 2 34 02 PM](https://github.com/zulip/zulip/assets/170719/21e3a290-701b-499c-a8b7-b0968306a167)
